core/src/main/java/org/elasticsearch/action/percolate/TransportPercolateAction.java

```
@@ -91,99 +91,136 @@ public void onFailure(Throwable e) {
+                XContentParser.Token token;
+                while ((token = parser.nextToken()) != XContentParser.Token.END_OBJECT) {
+                    if (token == XContentParser.Token.FIELD_NAME) {
+                        currentFieldName = parser.currentName();
+                    } else if (token == XContentParser.Token.START_OBJECT) {
+                        if ("doc".equals(currentFieldName)) {
+                            XContentBuilder builder = XContentFactory.jsonBuilder();
+                            builder.copyCurrentStructure(parser);
+                            builder.flush();
+                            documentSource = builder.bytes();
+                        } else if ("query".equals(currentFieldName) || "filter".equals(currentFieldName)) {
+                            XContentBuilder builder = XContentFactory.jsonBuilder();
+                            builder.copyCurrentStructure(parser);
+                            builder.flush();
+                            querySource = builder.bytes();
+                        } else if ("sort".equals(currentFieldName)) {
+                            searchSource.field("sort");
+                            searchSource.copyCurrentStructure(parser);
+                        } else if ("aggregations".equals(currentFieldName)) {
+                            searchSource.field("aggregations");
+                            searchSource.copyCurrentStructure(parser);
+                        } else if ("highlight".equals(currentFieldName)) {
+                            searchSource.field("highlight");
+                            searchSource.copyCurrentStructure(parser);
+                        }
+                    } else if (token == XContentParser.Token.START_ARRAY) {
+                        searchSource.field("sort");
+                        searchSource.copyCurrentStructure(parser);
+                    } else if (token == null) {
+                        break;
```

this looks like a hack to me, does it happen in practice? Should we throw an exception if it happens?